### PR TITLE
Fixes flaky RobotState test

### LINKED ISSUE #3103: `OneRobot.rigidlyAttachedParent` test fails in jobs
### Description

The test mentioned in the title, defined in `moveit_core/robot_state/test/robot_state_test.cpp` has been frequently failing on Github jobs.

https://github.com/moveit/moveit2/actions/runs/11860308971/job/33055270697
https://github.com/moveit/moveit2/actions/runs/11860056590/job/33054441260
https://github.com/moveit/moveit2/actions/runs/11782535125/job/32817780351
https://github.com/moveit/moveit2/actions/runs/11781483295/job/32814401931

to list a few.

Most notably, this seems to consistently affect the SonarScan job on main, despite other jobs which (I assume run the same tests) consistently passing

### ROS Distro

Humble

### OS and version

Rolling Docker Image

### Source or binary build?

Source

### If binary, which release version?

N/A

### If source, which branch?

main

### Which RMW are you using?

FastRTPS

### Steps to Reproduce

Unclear. The test fails fairly consistently on various jobs. I've tried running the test locally 10000 times with the following command...

```
cd ~/ws_moveit/build/moveit_core/robot_state
./test_robot_state --gtest_filter="*rigid*" --gtest_repeat=10000 | grep "PASSED" | wc -l
```

Which consistently prints out 10000 passes

### Expected behavior

Since the test pass consistently locally, it's reasonable to expect them to do so on the jobs.

### Actual behavior

Many PRs have failed with the error given below.

### Backtrace or Console output

On the jobs:

```
    [ RUN      ] OneRobot.rigidlyConnectedParent
    Error:   Group 'base_with_bad_subgroups' has unsatisfied subgroups
             at line 304 in /home/runner/work/moveit2/moveit2/.work/upstream_ws/src/srdfdom/src/model.cpp
    [INFO] [1731689210.102607487] [moveit_1614370902.moveit.core.robot_model]: Loading robot model 'one_robot'...
    [WARN] [1731689210.102743753] [moveit_1614370902.moveit.core.robot_state]: Returning dirty link transforms
    test_robot_state: /home/runner/work/moveit2/moveit2/.work/target_ws/src/moveit2/moveit_core/robot_state/src/robot_state.cpp:1309: const Eigen::Isometry3d& moveit::core::RobotState::getFrameInfo(const std::string&, const moveit::core::LinkModel*&, bool&) const: Assertion `checkLinkTransforms()' failed.
```

Running locally:
```
Note: Google Test filter = *rigid*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from OneRobot
[ RUN      ] OneRobot.rigidlyConnectedParent
Error:   Group 'base_with_bad_subgroups' has unsatisfied subgroups
         at line 304 in /root/ws_moveit/src/srdfdom/src/model.cpp
[WARN] [1731700916.222734076] [moveit_85773106]: exception thrown while creating node for logging: failed to create guard condition: context argument is null, at ./src/rcl/guard_condition.c:65
[WARN] [1731700916.222763620] [moveit_85773106]: if rclcpp::init was not called, messages from this logger may be missing from /rosout
[INFO] [1731700916.222779245] [moveit_85773106.moveit.core.robot_model]: Loading robot model 'one_robot'...
[ERROR] [1731700916.222920470] [moveit_85773106.moveit.core.robot_state]: Unable to find link for frame 'no_object'
[ERROR] [1731700916.222934489] [moveit_85773106.moveit.core.robot_state]: Unable to find link for frame 'object/no_subframe'
[ERROR] [1731700916.222937242] [moveit_85773106.moveit.core.robot_state]: Unable to find link for frame ''
[ERROR] [1731700916.222948465] [moveit_85773106.moveit.core.robot_state]: Unable to find link for frame 'object/'
[ERROR] [1731700916.222960187] [moveit_85773106.moveit.core.robot_state]: Unable to find link for frame '/'
[       OK ] OneRobot.rigidlyConnectedParent (5 ms)
[----------] 1 test from OneRobot (5 ms total)

[----------] Global test environment tear-down
[==========] 1 test
```

Interestingly, the failure always comes after a `Returning dirty link transforms` message, which does not occur when the tests pass locally. Given the failure is related to invalid transforms, this might be related.

### PR DESCRIPTION
**Pull request by TSNoble**
### Description

Fixes #3103 by ensuring that the `dirty_link_transforms_` member is cleared before querying the `RobotState`

See #3102 for more details on the cause. 

TLDR: 

Setting the build type to Release causes tests to silently fail, whereas Debug does not (some tests rely on assert calls, which are enabled / disabled based on the presence of the `NDBUG` macro variable, which is set by CMake depending on the build type). 

See the SonarScan pipeline call to ros-industrial/industrial_ci@master, which sets DCMAKE_BUILD_TYPE to Debug (and where the tests fail), and the same call in the rolling-ci pipeline (where the tests pass), which sets the same variable to Release.

I've also added a comment (and demonstration of the fix, at least for this test) locally.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/moveit/moveit2/blob/main/MIGRATION.md) notes
- [x] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [x] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/moveit/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by TSNoble** [comment]
Without the changes (Debug):

```
root@d1aa334f4187:~/ws_moveit# MAKEFLAGS='-j1' colcon build --cmake-args -DBUILD_TESTING=ON -DCMAKE_BUILD_TYPE=Debug --executor sequential --packages-up-to moveit_core
...
root@d1aa334f4187:~/ws_moveit# cd build/moveit_core/robot_state/ && ./test_robot_state --gtest_filter="*rigid*"
Note: Google Test filter = *rigid*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from OneRobot
[ RUN      ] OneRobot.rigidlyConnectedParent
Error:   Group 'base_with_bad_subgroups' has unsatisfied subgroups
         at line 304 in /root/ws_moveit/src/srdfdom/src/model.cpp
[WARN] [1731708234.797945875] [moveit_2006049342]: exception thrown while creating node for logging: failed to create guard condition: context argument is null, at ./src/rcl/guard_condition.c:65
[WARN] [1731708234.797975187] [moveit_2006049342]: if rclcpp::init was not called, messages from this logger may be missing from /rosout
[INFO] [1731708234.798013942] [moveit_2006049342.moveit.core.robot_model]: Loading robot model 'one_robot'...
[WARN] [1731708234.800753550] [moveit_2006049342.moveit.core.robot_state]: Returning dirty link transforms
test_robot_state: /root/ws_moveit/src/moveit2/moveit_core/robot_state/src/robot_state.cpp:1309: const Eigen::Isometry3d& moveit::core::RobotState::getFrameInfo(const std::string&, const moveit::core::LinkModel*&, bool&) const: Assertion `checkLinkTransforms()' failed.
Aborted
```

With the changes (Debug):

```
root@d1aa334f4187:~/ws_moveit# MAKEFLAGS='-j1' colcon build --cmake-args -DBUILD_TESTING=ON -DCMAKE_BUILD_TYPE=Debug --executor sequential --packages-up-to moveit_core
...
root@d1aa334f4187:~/ws_moveit# cd build/moveit_core/robot_state/ && ./test_robot_state --gtest_filter="*rigid*"
Note: Google Test filter = *rigid*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from OneRobot
[ RUN      ] OneRobot.rigidlyConnectedParent
Error:   Group 'base_with_bad_subgroups' has unsatisfied subgroups
         at line 304 in /root/ws_moveit/src/srdfdom/src/model.cpp
[WARN] [1731708144.887481051] [moveit_3787422709]: exception thrown while creating node for logging: failed to create guard condition: context argument is null, at ./src/rcl/guard_condition.c:65
[WARN] [1731708144.887508333] [moveit_3787422709]: if rclcpp::init was not called, messages from this logger may be missing from /rosout
[INFO] [1731708144.887537809] [moveit_3787422709.moveit.core.robot_model]: Loading robot model 'one_robot'...
[ERROR] [1731708144.890794401] [moveit_3787422709.moveit.core.robot_state]: Unable to find link for frame 'no_object'
[ERROR] [1731708144.890822982] [moveit_3787422709.moveit.core.robot_state]: Unable to find link for frame 'object/no_subframe'
[ERROR] [1731708144.890838105] [moveit_3787422709.moveit.core.robot_state]: Unable to find link for frame ''
[ERROR] [1731708144.890841557] [moveit_3787422709.moveit.core.robot_state]: Unable to find link for frame 'object/'
[ERROR] [1731708144.890853957] [moveit_3787422709.moveit.core.robot_state]: Unable to find link for frame '/'
[       OK ] OneRobot.rigidlyConnectedParent (9 ms)
[----------] 1 test from OneRobot (9 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (9 ms total)
[  PASSED  ] 1 test.
```

**Review by sea-bass — APPROVED** [review]
Nice find!

**Comment by sea-bass** [comment]
Does this need backport?

**Comment by TSNoble** [comment]
@sea-bass Yeah, that'd be great, thanks!

**Comment by TSNoble** [comment]
@sea-bass Out of curiosity, are main PRs automatically merged into jazzy too? This one in particular feels like it'll affect all versions

**Comment by sea-bass** [comment]
> @sea-bass Out of curiosity, are main PRs automatically merged into jazzy too? This one in particular feels like it'll affect all versions

There is no `jazzy` branch right now, since it's compatible with `main`.

What will happen is that, at some point, there will likely be a breaking change in Rolling that prompts us to cut a different `jazzy` branch that diverges from `main`. But until then, one and the same!
